### PR TITLE
DI: Consider capture of `let`s to be a read-only use.

### PR DESCRIPTION
**Pull request by jckarter**
SILGen emits capture of an address-only `let` property as @inout_aliasable, but Sema enforces that the closure is not really allowed to modify the property, so DI can consider the capture to be a read. Fixes rdar://problem/40828667.

A more principled solution would be to treat `let` captures as @in_guaranteed (see #17047), but that unfortunately leads to breakage elsewhere in the optimizer.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please smoke test

File: lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp [hunk 1]
```diff
@@ -877,9 +877,30 @@ void ElementUseCollector::collectUses(SILValue Pointer, unsigned BaseEltNo) {
...
       case ParameterConvention::Indirect_InoutAliasable: {
+        // FIXME: The @inout_aliasable convention is used for indirect captures
+        // of both 'let' and 'var' variables. Using a more specific convention
+        // for 'let' properties like @in_guaranteed unfortunately exposes bugs
+        // elsewhere in the pipeline. A 'let' capture cannot really be mutated
+        // by the callee, and this is enforced by sema, so we can consider it
+        // a nonmutating use.
+        bool isLet = true;
+        
+        for (unsigned i = 0; i < TheMemory.NumElements; ++i) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp]: Can you use none_of here instead? i.e.:

```
if (llvm::none_of(indices(TheMemory.NumElements),
                            [&](unsigned i) { return !TheMemory.isElementLetProperty(i) })) {
  addElementUses(...);
  continue;
}
```

File: lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp [hunk 2]
```diff
@@ -555,8 +555,28 @@ bool ElementUseCollector::collectUses(SILValue Pointer, unsigned BaseEltNo) {
...
       case ParameterConvention::Indirect_InoutAliasable: {
+        // FIXME: The @inout_aliasable convention is used for indirect captures
+        // of both 'let' and 'var' variables. Using a more specific convention
+        // for 'let' properties like @in_guaranteed unfortunately exposes bugs
+        // elsewhere in the pipeline. A 'let' capture cannot really be mutated
+        // by the callee, and this is enforced by sema, so we can consider it
+        // a nonmutating use.
+        bool isLet = true;
+        
+        for (unsigned i = 0; i < TheMemory.NumElements; ++i) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: Same here.

Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: Also, please add a test case for PMO.

Comment (jckarter) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: What is PMO?

File: lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp (hunk 1 above)
Comment (jckarter) [reply lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp]: Thanks for the suggestion, but I don't find that easier to read.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: PMO == Predictable Memory Opts. Predictable Mem Opts and DI now have two different use collector impls. So if you only care about DI, just change the Ownership one. I am waiting for you to finish this to rename the Ownership version to not have the Ownership suffix.

If your thing doesn't care about predictable memory opts, then I wouldn't change this.

File: lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp]: I am not going to push on this... but I just wanted to point out that I was just bringing up this section from the LLVM style guide:

https://www.llvm.org/docs/CodingStandards.html#turn-predicate-loops-into-predicate-functions

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

File: lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp (hunk 2 above)
Comment (jckarter) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: You may want to wait to change the name until we wrap up the 4.2 branch, since that will make cherry-picking changes over more difficult and error-prone. Does PMO affect diagnostics?

Comment (jckarter) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: Furthermore, does the formerly-DIMemoryUseCollector still have a role in DI on the 4.2 branch?

File: lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp (hunk 1 above)
Comment (jckarter) [reply lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp]: Even that doesn't recommend using `none_of`, though, but splitting the loop out into a separate function,

File: lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: No PMO does not. It is only used for predictable memory operations. The idea was to split the two passes so that they could evolve independently. They had a bunch of non-shared code that made it hard to reason about.

I think that git is good enough to handle moves... but I can wait.

Comment (jckarter) [reply lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp]: My concern with the move here is that you'd be replacing a file with a different-but-similar-enough file that Git might mistake it for an in-place modification rather than a sequence of renames.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

File: test/SILOptimizer/definite_init_diagnostics.swift [hunk 3]
```diff
@@ -83,11 +83,6 @@ func test2() {
   }
   b4 = 7
   
-  let b5: Any
```
Comment (slavapestov) [inline test/SILOptimizer/definite_init_diagnostics.swift]: Why did you remove this test case?

Comment (jckarter) [reply test/SILOptimizer/definite_init_diagnostics.swift]: The change to Sema makes the DI test redundant, and furthermore, since the error here gets diagnosed in Sema, it prevents any other DI tests in this file from being reached.

**Comment by jckarter** [comment]
@swift-ci Please test OS X
